### PR TITLE
Support UD on hip08

### PR DESCRIPTION
**Pull request by Li-Weihang**
This patchset enables hip08 to support Unreliable Datagram.

### CODE REVIEW

File: providers/hns/hns_roce_u_hw_v1.c [hunk 1]
```diff
@@ -471,19 +470,20 @@ static int hns_roce_u_v1_post_send(struct ibv_qp *ibvqp, struct ibv_send_wr *wr,
...
 	for (nreq = 0; wr; ++nreq, wr = wr->next) {
 		if (hns_roce_wq_overflow(&qp->sq, nreq,
 					 to_hr_cq(qp->ibv_qp.send_cq))) {
 			ret = -1;
 			*bad_wr = wr;
 			goto out;
 		}
+
+		wqe_idx = (qp->sq.head + nreq) & (qp->sq.wqe_cnt - 1);
+
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v1.c]: I didn't understand your problem and solution, but a couple of lines above you have hns_roce_wq_overflow() function which supposes to protect from overflow.

Comment (rleon) [reply providers/hns/hns_roce_u_hw_v1.c]: plus the "a %= b" construction is the better way to write it.

File: providers/hns/hns_roce_u_hw_v2.c [hunk 2]
```diff
@@ -637,6 +637,109 @@ static void set_sge(struct hns_roce_v2_wqe_data_seg *dseg,
...
+		     (wr->send_flags & IBV_SEND_SIGNALED) ? 1 : 0);
+	roce_set_bit(ud_sq_wqe->rsv_opcode, UD_SQ_WQE_SE_S,
+		     (wr->send_flags & IBV_SEND_SOLICITED) ? 1 : 0);
+	roce_set_bit(ud_sq_wqe->rsv_opcode, UD_SQ_WQE_OWNER_S,
+		     ~(((qp->sq.head + nreq) >> qp->sq.shift) & 0x1));
+
+	switch (wr->opcode) {
+	case IBV_WR_SEND:
+		hr_op = HNS_ROCE_WQE_OP_SEND;
+		ud_sq_wqe->immtdata = 0;
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: You cleared it a second before.

File: providers/hns/hns_roce_u_hw_v2.c [hunk 3]
```diff
@@ -609,28 +609,194 @@ static int hns_roce_u_v2_arm_cq(struct ibv_cq *ibvcq, int solicited)
...
+		    struct hns_roce_qp *qp, struct ibv_send_wr *wr,
+		    struct hns_roce_sge_info *sge_info)
+{
+	int i;
+
+	sge_info->valid_num = 0;
+	sge_info->total_len = 0;
+
+	for (i = 0; i < wr->num_sge; i++)
+		if (likely(wr->sg_list[i].length)) {
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: success-oriented flow, please

if (unlikely(,,,))
 continue;

..

File: providers/hns/hns_roce_u_hw_v2.c [hunk 4]
```diff
@@ -609,28 +609,194 @@ static int hns_roce_u_v2_arm_cq(struct ibv_cq *ibvcq, int solicited)
...
+	if (wr->opcode == IBV_WR_ATOMIC_FETCH_AND_ADD ||
+	    wr->opcode == IBV_WR_ATOMIC_CMP_AND_SWP) {
+		dseg++;
+		set_atomic_seg((struct hns_roce_wqe_atomic_seg *)dseg, wr);
+		return 0;
+	}
+
+	if (wr->send_flags & IBV_SEND_INLINE) {
+		if (wr->opcode == IBV_WR_RDMA_READ) {
+			fprintf(stderr, "Not supported inline data!\n");
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: No fprints in libraries

File: providers/hns/hns_roce_u_hw_v2.c [hunk 5]
```diff
@@ -609,28 +609,194 @@ static int hns_roce_u_v2_arm_cq(struct ibv_cq *ibvcq, int solicited)
...
+			return -EINVAL;
+		}
+
+		if (sge_info->total_len > qp->max_inline_data) {
+			fprintf(stderr,
+				"Failed to inline, data len=%d, max inline len=%d!\n",
```
Comment (rleon) [inline providers/hns/hns_roce_u_hw_v2.c]: ditto

File: providers/hns/hns_roce_u.h [hunk 6]
```diff
@@ -68,6 +68,8 @@
 #define HNS_ROCE_TPTR_OFFSET		0x1000
 #define HNS_ROCE_STATIC_RATE		3 /* Gbps */
 
+#define	ETH_ALEN			6
```
[the remark this answers is not in the document]
Comment (Li-Weihang) [reply providers/hns/hns_roce_u.h]: OK, will modify, thank you.

File: providers/hns/hns_roce_u_hw_v2.c (hunk 2 above)
Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: OK, will remove it, thank you.

File: providers/hns/hns_roce_u_hw_v2.c (hunk 3 above)
Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: OK, thank you for your advice :)

File: providers/hns/hns_roce_u_hw_v2.c (hunk 4 above)
Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: Hi Leon,
Do you means that we should print informations only when debug in libraries?
I also saw some fprintf in other drivers, could you please give me some suggestions on how to tell the users when an error occurs?

File: providers/hns/hns_roce_u_hw_v1.c (hunk 1 above)
Comment (eason-liuyx) [reply providers/hns/hns_roce_u_hw_v1.c]: > I didn't understand your problem and solution, but a couple of lines above you have hns_roce_wq_overflow() function which supposes to protect from overflow.
Hi, Leon, thanks. You are right, the commit message is uncorrect, I will rewrite it.

Comment (eason-liuyx) [reply providers/hns/hns_roce_u_hw_v1.c]: > I didn't understand your problem and solution, but a couple of lines above you have hns_roce_wq_overflow() function which supposes to protect from overflow.

Hi, Leon, thanks. You are right, the commit message is uncorrect, I will rewrite it.

Comment (eason-liuyx) [reply providers/hns/hns_roce_u_hw_v1.c]: > plus the "a %= b" construction is the better way to write it.

Hi Leon, do you mean: 
wqe_idx = (qp->sq.head + nreq)  %  qp->sq.wqe_cnt;

File: providers/hns/hns_roce_u_hw_v2.c (hunk 4 above)
Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: Hi Leon,
Do you means that we should print informations only when debug in libraries? Or it's not reasonable to print message here in inline mode?
I saw some fprintf in other drivers, should I replace all printf/fprintf in hns with a macro that only show message when debug?

Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: > No fprints in libraries

Hi Leon,
Do you means that we should print informations only when debug in libraries? Or it's not reasonable to print message here in inline mode?
I saw some fprintf in other drivers, should I replace all printf/fprintf in hns with a macro that only show message when debug?

File: providers/hns/hns_roce_u_hw_v2.c (hunk 5 above)
Comment (Li-Weihang) [reply providers/hns/hns_roce_u_hw_v2.c]: > ditto

Hi Leon,
I have modified code as suggested and pushed a new version, could you please review it again? Thank you :)

**Pull request closed without merging**